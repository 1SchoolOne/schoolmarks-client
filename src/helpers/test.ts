import { MockedObject } from 'vitest'

import { axiosInstance } from '@api/axiosInstance'

vi.mock('../api/axiosInstance')
/**
 * Permet de mocker les requÃªtes axios.
 *
 * @example
 *
 * mockAxios.get.mockResolvedValue({ data: [{userId: 1, first_name: 'toto' }] })
 */
export const mockAxios = axiosInstance as MockedObject<typeof axiosInstance>
