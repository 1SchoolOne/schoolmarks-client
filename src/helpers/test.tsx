import { PropsWithChildren, ThemeProvider } from '@1schoolone/ui'
import { QueryClient, QueryClientProvider } from '@tanstack/react-query'
import { RenderOptions, render, render as renderReact } from '@testing-library/react'
import { ReactNode, useMemo } from 'react'
import { RouterProvider, createMemoryRouter } from 'react-router-dom'

import { IdentityProvider } from '@contexts'

const testingQueryClient = new QueryClient({
	defaultOptions: { queries: { staleTime: 0, retry: false } },
})

function AllProviders({ children }: PropsWithChildren) {
	const testingRouter = useMemo(
		() =>
			createMemoryRouter(
				[{ path: '/:classSessionId', element: <IdentityProvider>{children}</IdentityProvider> }],
				{ initialEntries: ['/123'] },
			),
		[children],
	)

	return (
		<ThemeProvider>
			<RouterProvider router={testingRouter} />
		</ThemeProvider>
	)
}

function renderWithQueryClient(
	client: QueryClient,
	ui: React.ReactElement,
	options?: RenderOptions,
): ReturnType<typeof renderReact> {
	const { rerender, ...result } = renderReact(
		<QueryClientProvider client={client}>{ui}</QueryClientProvider>,
		options,
	)

	return {
		...result,
		rerender: (rerenderUi: React.ReactNode) =>
			rerender(<QueryClientProvider client={client}>{rerenderUi}</QueryClientProvider>),
	}
}

/** This helper will render the React node only with the ThemeProvider. */
export function renderWithThemeProvider(
	elements: ReactNode,
	options?: Omit<RenderOptions, 'wrapper'>,
) {
	return render(elements, { ...options, wrapper: ThemeProvider })
}

/**
 * This helper will render the React node with all the following providers:
 *
 * - ThemeProvider
 * - RouterProvider
 * - QueryClientProvider
 */
export function renderWithProviders(elements: ReactNode, options?: RenderOptions) {
	return renderWithQueryClient(testingQueryClient, <AllProviders>{elements}</AllProviders>, options)
}
