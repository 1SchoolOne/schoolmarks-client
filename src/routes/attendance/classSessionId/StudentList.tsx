import { useQuery } from '@tanstack/react-query'
import { Col, Divider, Flex, List, Row, Tag, Typography } from 'antd'
import dayjs from 'dayjs'
import { useContext } from 'react'
import { useLoaderData, useParams } from 'react-router-dom'

import { axiosInstance } from '@api/axiosInstance'
import { getClassSessionQueryOptions } from '@api/classSessions'

import { IdentityContext } from '@contexts'

import { hasPermission } from '@utils/permissions'

import { classSessionloader } from '..'
import { GetUserByIdResponse } from '../../../types/api/users'

export function StudentList() {
	const initialData = useLoaderData() as Awaited<ReturnType<typeof classSessionloader>>
	const { user } = useContext(IdentityContext)
	const params = useParams()
	const canReadCheckinSessions = hasPermission(user, 'read', 'checkin_sessions')

	const classSessionQueryOptions = getClassSessionQueryOptions(params.classSessionId)

	const { data: classSession } = useQuery({
		...classSessionQueryOptions,
		initialData,
		enabled: typeof params.classSessionId === 'string',
	})

	const { data: students, isPending } = useQuery({
		queryKey: ['checkinStudents', classSession.checkin_session?.id],
		queryFn: async () => {
			const { data } = await axiosInstance.get<
				Array<{
					id: string
					checkin_session: string
					student: GetUserByIdResponse
					checked_in_at: string
					status: string
				}>
			>(`/attendance_records/?checkin_session_id=${classSession.checkin_session?.id}`)

			return data
		},
		select: (data) => {
			if (Array.isArray(data)) {
				return data
			}

			return [data]
		},
		refetchInterval: 2_000,
		enabled: !!classSession.checkin_session,
	})

	if (!canReadCheckinSessions) {
		return <></>
	}

	// TODO: remove inline styling
	return (
		<>
			<Col span={2}>
				<Divider type="vertical" style={{ height: '100%' }} />
			</Col>
			<Col span={11}>
				{classSession.checkin_session ? (
					<List
						className="student-list"
						itemLayout="horizontal"
						dataSource={students}
						loading={isPending}
						locale={{
							emptyText: "Aucun élève ne s'est enregistré pour le moment",
						}}
						header={
							<Row gutter={4} align="middle">
								<Col span={13}>
									<Typography.Text strong>Nom</Typography.Text>
								</Col>
								<Col span={5}>
									<Typography.Text strong>Arrivée</Typography.Text>
								</Col>
								<Col span={6}>
									<Typography.Text strong>Statut</Typography.Text>
								</Col>
							</Row>
						}
						renderItem={(item) => {
							const checkedInAt = dayjs(item.checked_in_at)
							const isAbsent = checkedInAt > dayjs(classSession.checkin_session?.closed_at)
							const isLate = checkedInAt > dayjs(classSession.checkin_session?.started_at)

							return (
								<List.Item className="student-list__item">
									<Row gutter={4} align="middle">
										<Col span={13}>
											<Typography.Text>
												{item.student?.first_name} {item.student?.last_name}
											</Typography.Text>
										</Col>
										<Col span={5}>
											<Typography.Text>{checkedInAt.format('HH:mm')}</Typography.Text>
										</Col>
										<Col span={6}>
											<Tag
												bordered={false}
												color={isAbsent ? 'error' : isLate ? 'warning' : 'success'}
											>
												{isAbsent ? 'Absent' : isLate ? 'En retard' : 'Présent'}
											</Tag>
										</Col>
									</Row>
								</List.Item>
							)
						}}
					/>
				) : (
					<Flex className="no-checkin-session-message" justify="center" align="center">
						<Typography.Text type="secondary">
							La liste s'actualisera une fois l'appel lancé
						</Typography.Text>
					</Flex>
				)}
			</Col>
		</>
	)
}
