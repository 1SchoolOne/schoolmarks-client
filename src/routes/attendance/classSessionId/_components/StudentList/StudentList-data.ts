import { useQuery } from '@tanstack/react-query'
import axios from 'axios'
import { useContext } from 'react'
import { useLoaderData, useParams } from 'react-router-dom'

import { classSessionloader } from '@routes/attendance'

import { AXIOS_DEFAULT_CONFIG } from '@api/axios'
import { getClassSessionQueryOptions } from '@api/classSessions'
import { AttendanceRecord } from '@apiSchema/attendanceRecords'
import { GetUserByIdResponse } from '@apiSchema/users'

import { IdentityContext } from '@contexts'

import { hasPermission } from '@utils/permissions'

export type Student = Omit<AttendanceRecord, 'student_detail' | 'student'> & {
	student: GetUserByIdResponse
}

export function useData() {
	const initialData = useLoaderData() as Awaited<ReturnType<typeof classSessionloader>>
	const { user } = useContext(IdentityContext)
	const params = useParams()
	const canReadCheckinSessions = hasPermission(user, 'read', 'checkin_sessions')

	const classSessionQueryOptions = getClassSessionQueryOptions(params.classSessionId)

	const { data: classSession } = useQuery({
		...classSessionQueryOptions,
		initialData,
		enabled: typeof params.classSessionId === 'string',
	})

	const { data: students, isPending } = useQuery({
		queryKey: ['checkinStudents', classSession.checkin_session?.id],
		queryFn: async () => {
			const { data } = await axios.get<Student[]>(
				`/attendance_records/?checkin_session_id=${classSession.checkin_session?.id}`,
				AXIOS_DEFAULT_CONFIG,
			)

			return data
		},
		select: (data) => {
			if (Array.isArray(data)) {
				return data
			}

			return [data]
		},
		refetchInterval: 2_000,
		initialData: [],
		enabled: !!classSession.checkin_session,
	})

	return {
		classSession,
		students,
		isStudentsPending: isPending,
		canReadCheckinSessions,
	}
}
