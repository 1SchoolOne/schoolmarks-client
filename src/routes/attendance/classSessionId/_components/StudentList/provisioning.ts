import dayjs from 'dayjs'

const dummyDate = dayjs('2025-01-01T00:00:00.000Z')

export const provisioning = {
	present: {
		students: [
			{
				student: {
					username: 'mock.user',
					first_name: 'Mock',
					last_name: 'User',
				},
				checkin_session: '1',
				checked_in_at: dummyDate.add(2, 'minutes').toISOString(),
				status: 'present',
			},
		],
		isCheckInSessionOpened: true,
		canReadCheckinSessions: true,
	},
	late: {
		students: [
			{
				student: {
					username: 'mock.user',
					first_name: 'Mock',
					last_name: 'User',
				},
				checkin_session: '1',
				checked_in_at: dummyDate.add(10, 'minutes').toISOString(),
				status: 'late',
			},
		],
		isCheckInSessionOpened: true,
		canReadCheckinSessions: true,
	},
	absent: {
		students: [
			{
				student: {
					username: 'mock.user',
					first_name: 'Mock',
					last_name: 'User',
				},
				checkin_session: '1',
				checked_in_at: dummyDate.add(15, 'minutes').toISOString(),
				status: 'absent',
			},
		],
		isCheckInSessionOpened: true,
		canReadCheckinSessions: true,
	},
	noReadPerms: {
		students: [
			{
				student: {
					username: 'mock.user',
					first_name: 'Mock',
					last_name: 'User',
				},
				checkin_session: '1',
				checked_in_at: dummyDate.add(15, 'minutes').toISOString(),
				status: 'absent',
			},
		],
		isCheckInSessionOpened: true,
		canReadCheckinSessions: false,
	},
	sessionNotOpened: {
		students: [
			{
				student: {
					username: 'mock.user',
					first_name: 'Mock',
					last_name: 'User',
				},
				checkin_session: '1',
				checked_in_at: dummyDate.add(15, 'minutes').toISOString(),
				status: 'absent',
			},
		],
		isCheckInSessionOpened: false,
		canReadCheckinSessions: true,
	},
}
