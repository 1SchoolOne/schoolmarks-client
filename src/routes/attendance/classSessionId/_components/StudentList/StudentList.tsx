import { List as AntdList, Col, Flex, Row, Tag, Typography } from 'antd'
import dayjs from 'dayjs'

import { Student, useData } from './StudentList-data'

export interface ListProps {
	canReadCheckinSessions: boolean
	isCheckInSessionOpened: boolean
	students: Student[]
	isLoading: boolean
}

export function StudentList() {
	const { canReadCheckinSessions, classSession, students, isStudentsPending } = useData()

	return (
		<List
			canReadCheckinSessions={canReadCheckinSessions}
			isCheckInSessionOpened={!!classSession.checkin_session}
			students={students}
			isLoading={isStudentsPending}
		/>
	)
}

export function List(props: ListProps) {
	const { canReadCheckinSessions, isCheckInSessionOpened, students, isLoading } = props

	if (!canReadCheckinSessions) {
		return <></>
	}

	return isCheckInSessionOpened ? (
		<AntdList
			className="student-list"
			itemLayout="horizontal"
			dataSource={students}
			loading={isLoading}
			locale={{
				emptyText: "Aucun élève ne s'est enregistré pour le moment",
			}}
			header={
				<Row gutter={4} align="middle">
					<Col span={13}>
						<Typography.Text strong>Nom</Typography.Text>
					</Col>
					<Col span={5}>
						<Typography.Text strong>Arrivée</Typography.Text>
					</Col>
					<Col span={6}>
						<Typography.Text strong>Statut</Typography.Text>
					</Col>
				</Row>
			}
			renderItem={(item) => {
				const checkedInAt = dayjs(item.checked_in_at)
				const isAbsent = item.status === 'absent'
				const isLate = item.status === 'late'

				return (
					<AntdList.Item className="student-list__item">
						<Row gutter={4} align="middle">
							<Col span={13}>
								<Typography.Text>
									{item.student?.first_name} {item.student?.last_name}
								</Typography.Text>
							</Col>
							<Col span={5}>
								<Typography.Text>{checkedInAt.format('HH:mm')}</Typography.Text>
							</Col>
							<Col span={6}>
								<Tag bordered={false} color={isAbsent ? 'error' : isLate ? 'warning' : 'success'}>
									{isAbsent ? 'Absent' : isLate ? 'En retard' : 'Présent'}
								</Tag>
							</Col>
						</Row>
					</AntdList.Item>
				)
			}}
		/>
	) : (
		<Flex className="no-checkin-session-message" justify="center" align="center">
			<Typography.Text type="secondary">
				La liste s'actualisera une fois l'appel lancé
			</Typography.Text>
		</Flex>
	)
}
