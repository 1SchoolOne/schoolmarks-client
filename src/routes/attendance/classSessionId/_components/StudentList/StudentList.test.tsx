import { screen } from '@testing-library/react'

import { renderWithThemeProvider } from '@helpers/test'

import { List } from './StudentList'
import { provisioning } from './provisioning'

const { present, late, absent, noReadPerms, sessionNotOpened } = provisioning

describe('<StudentList />', () => {
	it('conforme au snapshot', () => {
		const { asFragment } = renderWithThemeProvider(
			<List
				canReadCheckinSessions
				isCheckInSessionOpened
				students={present.students}
				isLoading={false}
			/>,
		)

		expect(asFragment()).toMatchSnapshot()
	})

	it('affiche les étudiants présents', () => {
		renderWithThemeProvider(<List isLoading={false} {...present} />)

		expect(screen.getByText('Mock User')).toBeDefined()
		expect(screen.getByText('Présent')).toBeDefined()
	})

	it('affiche les étudiants en retard', () => {
		renderWithThemeProvider(<List isLoading={false} {...late} />)

		expect(screen.getByText('Mock User')).toBeDefined()
		expect(screen.getByText('En retard')).toBeDefined()
	})

	it('affiche les étudiants absents', () => {
		renderWithThemeProvider(<List isLoading={false} {...absent} />)

		expect(screen.getByText('Mock User')).toBeDefined()
		expect(screen.getByText('Absent')).toBeDefined()
	})

	it("n'affiche rien lors que l'utilisateur n'a pas les permissions", () => {
		renderWithThemeProvider(<List isLoading={false} {...noReadPerms} />)

		expect(screen.queryByText('Mock User')).toBeNull()
		expect(screen.queryByText('Absent')).toBeNull()
	})

	it("affiche un message lorsque la session d'appel n'est pas encore ouverte", () => {
		renderWithThemeProvider(<List isLoading={false} {...sessionNotOpened} />)

		expect(screen.queryByText('Mock User')).toBeNull()
		expect(screen.queryByText('Absent')).toBeNull()

		expect(screen.getByText("La liste s'actualisera une fois l'appel lancé")).toBeDefined()
	})
})
