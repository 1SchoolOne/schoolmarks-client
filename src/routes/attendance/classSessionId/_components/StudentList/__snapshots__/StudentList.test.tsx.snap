// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<StudentList /> > conforme au snapshot 1`] = `
<DocumentFragment>
  <div
    class="ant-list ant-list-split student-list css-dev-only-do-not-override-yta69t css-var-r0"
  >
    <div
      class="ant-list-header"
    >
      <div
        class="ant-row ant-row-middle css-dev-only-do-not-override-yta69t css-var-r0"
        style="margin-left: -2px; margin-right: -2px;"
      >
        <div
          class="ant-col ant-col-13 css-dev-only-do-not-override-yta69t css-var-r0"
          style="padding-left: 2px; padding-right: 2px;"
        >
          <span
            class="ant-typography css-dev-only-do-not-override-yta69t css-var-r0"
          >
            <strong>
              Nom
            </strong>
          </span>
        </div>
        <div
          class="ant-col ant-col-5 css-dev-only-do-not-override-yta69t css-var-r0"
          style="padding-left: 2px; padding-right: 2px;"
        >
          <span
            class="ant-typography css-dev-only-do-not-override-yta69t css-var-r0"
          >
            <strong>
              Arrivée
            </strong>
          </span>
        </div>
        <div
          class="ant-col ant-col-6 css-dev-only-do-not-override-yta69t css-var-r0"
          style="padding-left: 2px; padding-right: 2px;"
        >
          <span
            class="ant-typography css-dev-only-do-not-override-yta69t css-var-r0"
          >
            <strong>
              Statut
            </strong>
          </span>
        </div>
      </div>
    </div>
    <div
      class="ant-spin-nested-loading css-dev-only-do-not-override-yta69t css-var-r0"
    >
      <div
        class="ant-spin-container"
      >
        <ul
          class="ant-list-items"
        >
          <li
            class="ant-list-item student-list__item"
          >
            <div
              class="ant-row ant-row-middle css-dev-only-do-not-override-yta69t css-var-r0"
              style="margin-left: -2px; margin-right: -2px;"
            >
              <div
                class="ant-col ant-col-13 css-dev-only-do-not-override-yta69t css-var-r0"
                style="padding-left: 2px; padding-right: 2px;"
              >
                <span
                  class="ant-typography css-dev-only-do-not-override-yta69t css-var-r0"
                >
                  Mock User
                </span>
              </div>
              <div
                class="ant-col ant-col-5 css-dev-only-do-not-override-yta69t css-var-r0"
                style="padding-left: 2px; padding-right: 2px;"
              >
                <span
                  class="ant-typography css-dev-only-do-not-override-yta69t css-var-r0"
                >
                  00:02
                </span>
              </div>
              <div
                class="ant-col ant-col-6 css-dev-only-do-not-override-yta69t css-var-r0"
                style="padding-left: 2px; padding-right: 2px;"
              >
                <span
                  class="ant-tag ant-tag-success ant-tag-borderless css-dev-only-do-not-override-yta69t css-var-r0"
                >
                  Présent
                </span>
              </div>
            </div>
          </li>
        </ul>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
